cmd/record/record.go

```diff
@@ -0,0 +1,217 @@
+		return 0, fmt.Errorf("unsupported protocol: %s", proto)
+	}
+}
+
+func parseFilters(filters []string) ([]*recorderpb.Filter, error) {
+	if len(filters) == 0 {
+		return nil, errors.New("need to specify at least one filter")
+	}
+
+	result := make([]*recorderpb.Filter, 0, len(filters))
+	for _, filter := range filters {
+		f := strings.SplitN(filter, fitlerSep, 5)
+		if len(f) != 5 {
+			return nil, fmt.Errorf("invalid 5-tuple filter: %q", f)
+		}
+
+		srcPrefix := f[0]
+		srcPort, err := strconv.ParseUint(f[1], 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("invalid source port: %s", f[1])
+		}
+		dstPrefix := f[2]
+		dstPort, err := strconv.ParseUint(f[3], 10, 16)
+		if err != nil {
+			return nil, fmt.Errorf("invalid destination port: %s", f[3])
+		}
+		proto, err := parseProto(f[4])
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, &recorderpb.Filter{
+			SourceCidr:      srcPrefix,
+			SourcePort:      uint32(srcPort),
+			DestinationCidr: dstPrefix,
+			DestinationPort: uint32(dstPort),
+			Protocol:        proto,
+		})
+	}
+
+	return result, nil
+}
+
+func fmtTime(resp *recorderpb.RecordResponse) string {
+	return resp.GetTime().AsTime().Format(time.RFC3339)
+}
+
+func fmtStats(stats *recorderpb.RecordingStatistics) string {
+	output := fmt.Sprintf("%d packets (%d bytes) written",
+		stats.GetPacketsCaptured(), stats.GetBytesCaptured())
+	if stats.GetPacketsLost() > 0 || stats.GetBytesLost() > 0 {
+		output = fmt.Sprintf("%s and %d packets (%d bytes) lost",
+			output, stats.GetPacketsLost(), stats.GetBytesLost())
+	}
+	return output
+}
+
+func runRecord(ctx context.Context, conn *grpc.ClientConn, filters []*recorderpb.Filter) error {
+	isTTY := false
+	if fileInfo, _ := os.Stdout.Stat(); (fileInfo.Mode() & os.ModeCharDevice) != 0 {
+		isTTY = true
+	}
+
+	recorder, err := recorderpb.NewRecorderClient(conn).Record(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = recorder.Send(&recorderpb.RecordRequest{
+		RequestType: &recorderpb.RecordRequest_Start{
+			Start: &recorderpb.StartRecording{
+				Filesink: &recorderpb.FileSinkConfiguration{
+					FilePrefix: fileSinkPrefix,
+				},
+				Include:          filters,
+				MaxCaptureLength: maxCaptureLength,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "Started recording. Press CTRL+C to stop.\n")
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		fmt.Fprintf(os.Stderr, "Stopping recording...\n")
+		err := recorder.Send(&recorderpb.RecordRequest{
+			RequestType: &recorderpb.RecordRequest_Stop{
+				Stop: &recorderpb.StopRecording{},
+			},
+		})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to to send stop recording message: %s\n", err)
+		}
+	}()
+
+	statusPrinted := false
+	for {
+		stats, err := recorder.Recv()
+		switch {
+		case err != nil:
+			return err
+		case stats.GetRunning() != nil:
+			if isTTY && statusPrinted {
+				fmt.Print(resetLastLine)
```

Comment:
not clear to me what this does, exactly

Reply:
This clears/resets the current line in stdout if stdout is a terminal similar to how tcpdump works. Probably best explained via video:

https://user-images.githubusercontent.com/50564/115267955-0fc9f180-a13a-11eb-9260-1255623277a6.mp4